### ISSUE TITLE
Inline list literals are widened to list[str], but equivalent named variables are not

### ISSUE THREAD

**Issue by ethancowardhive** [astral-sh/ty#3065]
### Summary

`list[Literal["..."]]` and `list[LiteralString]` are rejected where `list[str]` is expected, despite the elements being strings. Inline list literals are accepted in the same position, suggesting the widening logic is being applied inconsistently depending on how the value is expressed.

https://play.ty.dev/d26e84a2-9d31-492e-8102-b98f1727ca9b

```python
from typing import Literal, reveal_type

s: Literal["hello"] = "hello"
t = s.title()
reveal_type(s) # Literal["hello"]
reveal_type(t) # LiteralString


def f(xs: list[str]) -> None: ...


# no error - ty widens the list literal
f([s])
named = [s]
# error - Expected `list[str]`, found `list[Literal["hello"]]`
f(named)

# no error - ty widens the list literal
f([t])
named = [t]
# error - Expected `list[str]`, found `list[LiteralString]`
f(named)
```

### Version

ty 0.0.23 (be6e94a27 2026-03-13)

**Comment by AlexWaygood** [comment]
Thanks for the report. For what it's worth, ty's behaviour matches mypy, pyrefly and zuban here -- the only major type checker that accepts this snippet without any complaints is pyright.

The problem here is that without a type annotation for your `named` variable, there is an ambiguity introduced as to whether you would like us to infer `named` as being `list[str]` or `list[Literal["hello"]]`. Either is an equally valid way to solve the type variable -- the question is, which is the *best* way for ty to solve the type variable in this scenario? When the variable is inline, the answer is clear: the variable is immediately passed to a parameter that expects `list[str]`, so we use that as type context to inform how we solve the type variable. With the `named` variable, however, there is no *immediate* type context to help us inform our inference.

In this scenario, we would usually default to a heuristic where we widen `Literal` types to instance types when solving type variables for invariant containers like `list`. This heuristic reduces false positives on user code. But in your _specific_ scenario, we don't do that -- and this is because you explicitly annotated `s` with `Literal["hello"]` earlier on in your code. We track that and assume that it means that you probably don't want us to "promote" `s` to an instance type anywhere in that scope, even in an invariant context like this.

Now, there is still more type context here that we could take account of but don't! While there is no _immediate_ type context here like there is when a list is passed _immediately_ to a parameter, there is the fact that we can see that `named` is _later on in the scope_ passed to a parameter annotated with `list[str]`. We intend to soon start taking account of type context "at a distance" like this when solving type variables -- this is tracked in https://github.com/astral-sh/ty/issues/1473.

For now, I would advise giving your `named` variable a type annotation, which will give ty _immediate_ type context to use when inferring a type for `named`, and will resolve the false positives on your code.
